thewalrus/fock_gradients.py

```diff
@@ -0,0 +1,270 @@
+    for n in range(cutoff):
+        for m in range(cutoff):
+            gradT[n, m] = 1j * np.sqrt(m + 1) * T[n, m + 1] * pref
+            if m > 0:
+                gradT[n, m] += 1j * np.sqrt(m) * T[n, m - 1] * pref
+    return gradT
+
+
+def Zgate(p, cutoff, grad=False, hbar=2, r=np.arcsinh(1.0)):
+    r"""
+    Calculates the Fock representation of the Zgate and its gradient
+    Arg:
+        p (float): Parameter of the gate
+        cutoff (int): Fock ladder cutoff
+        grad (boolean): Whether to calculate the gradient or not
+        hbar (float): Value of hbar is the commutation relation
+        r (float): Value of the parameter used internally in fock_tensor.
+    Returns:
+        tuple(G,dG): The Fock representations of the gate and its gradient
+    """
+    nmodes = 1
+    pref = 1.0 / np.sqrt(2 * hbar)
+    alpha = np.array([1j * pref * p])
+    S = np.identity(2 * nmodes)
+
+    if not grad:
+        return fock_tensor(S, alpha, cutoff, r=r), None
+
+    T = fock_tensor(S, alpha, cutoff + 1, r=r)
+    return T[0:cutoff, 0:cutoff], grad_Zgate(T, pref)
+
+
+@jit("double[:,:](double[:,:])")
+def grad_Sgate(T):
+    """Calculate the gradient of the Sgate
+    Args:
+        T (array): Array representing the gate
+    Returns:
+        (array): Array representing the gradient of the gate
+    """
+    cutoff, _ = T.shape
+    cutoff -= 2
+    gradT = np.zeros([cutoff, cutoff])
+    for n in range(cutoff):
+        offset = n % 2
+        for m in range(offset, cutoff, 2):
+            gradT[n, m] = -0.5 * np.sqrt((m + 1) * (m + 2)) * T[n, m + 2]
+            if m > 1:
+                gradT[n, m] += 0.5 * np.sqrt(m * (m - 1)) * T[n, m - 2]
+
+    return gradT
+
+
+def Sgate(s, cutoff, grad=False, r=np.arcsinh(1.0)):
+    r"""
+    Calculates the Fock representation of the Sgate and its gradient
+    Arg:
+        s (float): Parameter of the gate
+        cutoff (int): Fock ladder cutoff
+        grad (boolean): Whether to calculate the gradient or not
+        r (float): Value of the parameter used internally in fock_tensor.
+    Returns:
+        tuple(G,dG): The Fock representations of the gate and its gradient
+    """
+    S = squeezing(s, 0.0)
+    if not grad:
+        return fock_tensor(S, np.zeros([1]), cutoff, r=r).real, None
+
+    T = fock_tensor(S, np.zeros([1]), cutoff + 2, r=r).real
+    return T[0:cutoff, 0:cutoff], grad_Sgate(T)
+
+
+def Rgate(theta, cutoff, grad=False):
+    r"""
+    Calculates the Fock representation of the Rgate and its gradient
+    Arg:
+        theta (float): Parameter of the gate
+        cutoff (int): Fock ladder cutoff
+        grad (boolean): Whether to calculate the gradient or not
+    Returns:
+        tuple(G,dG): The Fock representations of the gate and its gradient
+    """
+    ns = np.arange(cutoff)
+    T = np.exp(1j * theta) ** ns
+    if not grad:
+        return np.diag(T), None
+    return np.diag(T), np.diag(1j * ns * T)
```

Comment:
Here the returned shape of both the tensor and the grad tensor are different again from previous cases